File: utils/date_format.py
from __future__ import annotations

import re
from datetime import date, datetime

DISPLAY_DATE_FORMAT = "%d/%m/%Y"
_ORDINAL_DATE = re.compile(r"^(\d{1,2})(?:st|nd|rd|th)?\s+([A-Za-z]+),?\s+(\d{4})$", re.IGNORECASE)


def format_display_date(
    value: date | datetime | str | None,
    fallback: str = "—",
) -> str:
    """Format dates for user-facing output as dd/mm/yyyy."""
    if value is None:
        return fallback

    if isinstance(value, datetime):
        return value.strftime(DISPLAY_DATE_FORMAT)

    if isinstance(value, date):
        return value.strftime(DISPLAY_DATE_FORMAT)

    text = str(value).strip()
    if not text:
        return fallback

    if len(text) >= 10 and text[2] == "/" and text[5] == "/":
        return text[:10]

    match = _ORDINAL_DATE.match(text)
    if match:
        day, month_name, year = match.groups()
        try:
            parsed = datetime.strptime(f"{day} {month_name} {year}", "%d %B %Y")
            return parsed.strftime(DISPLAY_DATE_FORMAT)
        except ValueError:
            pass

    try:
        return date.fromisoformat(text[:10]).strftime(DISPLAY_DATE_FORMAT)
    except ValueError:
        return text

File: frontend/src/utils/dateFormat.ts
const ISO_DATE_PREFIX = /^(\d{4})-(\d{2})-(\d{2})/;
const DISPLAY_DATE = /^\d{2}\/\d{2}\/\d{4}$/;

/** Format API/ISO dates for display as dd/mm/yyyy. */
export function formatDisplayDate(
  value: string | null | undefined,
  fallback = "—",
): string {
  if (value == null) return fallback;

  const text = String(value).trim();
  if (!text) return fallback;
  if (DISPLAY_DATE.test(text)) return text;

  const isoMatch = text.match(ISO_DATE_PREFIX);
  if (isoMatch) {
    const [, year, month, day] = isoMatch;
    return `${day}/${month}/${year}`;
  }

  return text;
}
